### PR TITLE
Audit log Maintenance Automation test

### PR DESCRIPTION
**Pull request by labkey-sweta**
#### Rationale
Automation test for Audit log cleanup

https://docs.google.com/document/d/1c4Fcbth-IW1msqhqzO8jwxik3ZwfTEfKL-uSN0n2sAc/edit#
#### Related Pull Requests
https://github.com/LabKey/professional/pull/19

#### Changes
*

### CODE REVIEW

File: src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+public class AuditLogMaintenancePage extends LabKeyPage<AuditLogMaintenancePage.ElementCache>
+{
+    public AuditLogMaintenancePage(WebDriver driver)
+    {
+        super(driver);
+        waitForPage();
```
Comment (labkey-tchad) [inline src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java]: The `LabKeyPage` default constructor calls `waitForPage`, no need to call it again here.
```suggestion
        super(driver);
```

File: src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    public void clickSave()
+    {
+        elementCache().saveBtn.click();
+    }
+
+    public void clickCancel()
+    {
+        elementCache().cancelBtn.click();
+    }
```
Comment (labkey-tchad) [inline src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java]: Do save and cancel trigger page loads?
```suggestion
    public void clickSave()
    {
        clickAndWait(elementCache().saveBtn);
    }

    public void clickCancel()
    {
        clickAndWait(elementCache().cancelBtn);
    }
```

File: src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    protected AuditLogMaintenancePage.ElementCache newElementCache()
+    {
+        return new AuditLogMaintenancePage.ElementCache();
+    }
+
+    public String getRetentionTime()
+    {
+        return elementCache().retentionTime.getText();
+    }
```
Comment (labkey-tchad) [inline src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java]: `getText` doesn't work like this for `select` elements. Need to use a `Select` or `SelectWrapper` component. (See note in `ElementCache`)

File: src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    public void clickCancel()
+    {
+        elementCache().cancelBtn.click();
+    }
+
+    protected class ElementCache extends LabKeyPage.ElementCache
+    {
+        WebElement retentionTime = Locator.name("retentionTime").findWhenNeeded(this);
```
Comment (labkey-tchad) [inline src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java]: Use component for select.
```suggestion
        Select retentionTime = SelectWrapper.Select(Locator.name("retentionTime")).findWhenNeeded(this);
```

File: src/org/labkey/test/tests/AuditLogMaintenanceTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        auditLogMaintenancePage.setRetentionTime("0 seconds - all audit log data will be deleted! For testing purposes only!")
+                .setExportDataBeforeDeleting(true)
+                .clickSave();
+
+        log("Running audit log maintenance job");
+        goToAdminConsole().clickSystemMaintenance();
+        waitAndClick(Locator.linkWithText("Audit Log Maintenance"));
+        switchToWindow(1);
+        shortWait().until(ExpectedConditions.textToBePresentInElementLocated(Locator.id("status-text"), "COMPLETE"));
+        switchToMainWindow();
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/AuditLogMaintenanceTest.java]: There's an existing test page for pipeline status (`PipelineStatusDetailsPage`).
```suggestion
        switchToWindow(1);
        new PipelineStatusDetailsPage(getDriver()).waitForStatus("COMPLETE");
        switchToMainWindow();
```

File: src/org/labkey/test/tests/AuditLogMaintenanceTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        log("Verifying audit logs got deleted");
+        ShowAuditLogPage auditLogPage = goToAdminConsole().clickAuditLog();
+        auditLogPage.selectView("List events");
+        checker().verifyTrue("Audit log table not cleared", auditLogPage.getLogTable().getDataRowCount() == 0);
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/AuditLogMaintenanceTest.java]: Just use a normal assertion here. The `checker` gets cranky when you don't take a screenshot after a failure.
```suggestion
        Assert.assertTrue("Audit log table not cleared", auditLogPage.getLogTable().getDataRowCount() == 0);
```

File: src/org/labkey/test/tests/AuditLogMaintenanceTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @After
+    public void restoreSettings()
+    {
+        goToAdminConsole().clickAuditLogMaintenance()
+                .setRetentionTime("None - all audit log data is retained indefinitely")
+                .setExportDataBeforeDeleting(false)
+                .clickSave();
+    }
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/AuditLogMaintenanceTest.java]: Don't use the UI in `@After` and `@AfterClass` methods. It interferes with test failure handling and screenshots. This should be easy enough to do via API.
```suggestion
    @After
    public void restoreSettings()
    {
        SimplePostCommand command = new SimplePostCommand("professional", "auditLogMaintenance");
        command.setJSONObject(new JSONObject(Map.of("retentionTime", -1, "export", false)));
        command.execute(createDefaultConnection(), null);
    }
```

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
Looks like this functionality comes from the `professional` module. This test and page belong in that repository.

File: src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java (hunk 1 above)
Comment (labkey-sweta) [reply src/org/labkey/test/pages/core/admin/AuditLogMaintenancePage.java]: Done

**Review by labkey-adam — APPROVED** [review]

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
`testAutomation` shouldn't reference classes in other repositories.
`ShowAdminPage` doesn't need to know about all the pages it might have links to. Modules can register all sorts of admin actions.

**Review by labkey-tchad — APPROVED** [review]

**Pull request merged**